// storybook-design-token/dist/components/primitives/Card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var theming_1 = require("@storybook/theming");
var usePanelWidth_1 = require("../../hooks/usePanelWidth");
var Container = theming_1.styled.div(function (_a) {
    var panelWidth = _a.panelWidth;
    return ({
        border: '1px solid #f1f1f1',
        borderRadius: '4px',
        display: 'flex',
        flexBasis: panelWidth === 'narrow'
            ? 'calc(100% - 8px)'
            : panelWidth === 'wide'
                ? 'calc(50% - 8px)'
                : 'calc(25% - 8px)',
        flexDirection: 'column',
        marginRight: '4px',
        marginLeft: '4px',
        minWidth: 0
    });
});
var Header = theming_1.styled.div(function () { return ({
    backgroundColor: '#343434',
    borderTopLeftRadius: '4px',
    borderTopRightRadius: '4px',
    color: '#fff',
    fontWeight: 500,
    padding: '8px'
}); });
var Body = theming_1.styled.div(function () { return ({}); });
var Section = theming_1.styled.div(function () { return ({
    padding: '8px',
    '& > strong': {
        display: 'block',
        marginBottom: '4px'
    }
}); });
var Preview = theming_1.styled.div(function () { return ({
    border: '1px solid #f1f1f1',
    margin: '8px',
    overflow: 'hidden',
    padding: '2px'
}); });
var Value = theming_1.styled.div(function () { return ({
    fontFamily: 'monospace'
}); });
exports.Card = function (_a) {
    var aliases = _a.aliases, description = _a.description, _b = _a.descriptionLabel, descriptionLabel = _b === void 0 ? 'Description' : _b, preview = _a.preview, title = _a.title, value = _a.value;
    var panelWidth = usePanelWidth_1.usePanelWidth();
    return (React.createElement(Container, { panelWidth: panelWidth },
        React.createElement(Header, null, title),
        React.createElement(Body, null,
            preview && React.createElement(Preview, null, preview),
            React.createElement(Section, null,
                React.createElement("strong", null, "Value"),
                React.createElement(Value, null, value)),
            description && (React.createElement(Section, null,
                React.createElement("strong", null, descriptionLabel),
                description)),
            aliases && (React.createElement(Section, null,
                React.createElement("strong", null, "Aliases"),
                aliases)))));
};
