// storybook-design-token/dist/components/primitives/Input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var theming_1 = require("@storybook/theming");
var Icons_1 = require("./Icons");
var Container = theming_1.styled.div(function () { return ({
    display: 'flex',
    '& > input': {
        display: 'block',
        fontFamily: 'monospace',
        padding: '4px',
        width: '100%'
    },
    '& > button': {
        alignItems: 'center',
        backgroundColor: '#f1f1f1',
        border: 'none',
        borderRadius: '2px',
        cursor: 'pointer',
        display: 'flex',
        height: '28px',
        justifyContent: 'center',
        padding: 0,
        marginLeft: '4px',
        width: '28px',
        '&:hover': {
            backgroundColor: '#f1f1f1'
        },
        '& > svg': {
            height: '16px',
            width: '16px'
        }
    }
}); });
exports.Input = function (_a) {
    var onChange = _a.onChange, onReset = _a.onReset, showReset = _a.showReset, value = _a.value;
    return (React.createElement(Container, null,
        React.createElement("input", { onChange: function (event) { return onChange(event.target.value); }, type: "text", value: value }),
        showReset && (React.createElement("button", { onClick: function () { return onReset(); }, type: "button" }, Icons_1.IconClose))));
};
