// storybook-design-token/dist/components/primitives/TokenName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_use_clipboard_1 = require("react-use-clipboard");
var theming_1 = require("@storybook/theming");
var Container = theming_1.styled.div(function () { return ({
    alignItems: 'center',
    cursor: 'pointer',
    display: 'flex',
    '& > span': {
        fontSize: '10px',
        fontWeight: 400,
        paddingLeft: '8px',
        whiteSpace: 'nowrap'
    }
}); });
var Name = theming_1.styled.div(function () { return ({
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap'
}); });
exports.TokenName = function (_a) {
    var token = _a.token;
    var _b = React.useState(false), showCopiedLabel = _b[0], setShowCopiedLabel = _b[1];
    var _c = react_use_clipboard_1.default(token.key), isCopied = _c[0], setCopied = _c[1];
    React.useEffect(function () {
        var timer;
        if (showCopiedLabel) {
            timer = setTimeout(function () {
                setShowCopiedLabel(false);
            }, 3000);
        }
        return function () { return clearTimeout(timer); };
    }, [showCopiedLabel]);
    return (React.createElement(Container, { onClick: function () {
            setCopied();
            setShowCopiedLabel(true);
        } },
        React.createElement(Name, null, token.key),
        React.createElement("span", { className: "copy-notice" }, showCopiedLabel ? 'Copied!' : 'Click to copy')));
};
