// storybook-design-token/dist/components/presenter/EasingToken.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var core_1 = require("@emotion/core");
var styled_1 = require("@emotion/styled");
var slide = core_1.keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  from {\n    transform: translate3d(0, 0, 0);\n  }\n  to {\n    transform: translate3d(114px, 0, 0);\n  }\n"], ["\n  from {\n    transform: translate3d(0, 0, 0);\n  }\n  to {\n    transform: translate3d(114px, 0, 0);\n  }\n"])));
var SmallBox = styled_1.default.div(function () { return ({
    position: 'relative',
    width: '25px',
    height: '25px',
    borderRadius: '4px',
    backgroundColor: '#222',
    animation: slide + " 1s ease infinite alternate",
    zIndex: 2
}); });
exports.EasingToken = function (_a) {
    var easing = _a.easing;
    return React.createElement(SmallBox, { style: { animationTimingFunction: easing } });
};
var templateObject_1;
