// react-datocms/dist/cjs/Seo/__tests__/index.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
var React = __importStar(require("react"));
var enzyme_1 = require("enzyme");
var index_1 = require("../index");
var metaTags = [
    {
        attributes: null,
        content: "A new Media Area is online! - DatoCMS",
        tag: "title"
    },
    {
        attributes: {
            property: "og:title",
            content: "A new Media Area is online!"
        },
        content: null,
        tag: "meta"
    },
    {
        attributes: {
            name: "twitter:title",
            content: "A new Media Area is online!"
        },
        content: null,
        tag: "meta"
    },
    {
        attributes: {
            sizes: "16x16",
            type: "image/png",
            rel: "icon",
            href: "https://example.org/favicon.png?h=16&w=16"
        },
        content: null,
        tag: "link"
    },
    {
        attributes: {
            sizes: "32x32",
            type: "image/png",
            rel: "icon",
            href: "https://example.org/favicon.png?h=32&w=32"
        },
        content: null,
        tag: "link"
    }
];
describe("renderMetaTags", function () {
    it("generates an array of meta tags", function () {
        var wrapper = enzyme_1.shallow(React.createElement("head", null, index_1.renderMetaTags(metaTags)));
        expect(wrapper).toMatchSnapshot();
    });
});
describe("renderMetaTagsToString", function () {
    it("generates an array of meta tags", function () {
        var result = index_1.renderMetaTagsToString(metaTags);
        expect(result).toEqual([
            '<title>A new Media Area is online! - DatoCMS</title>',
            '<meta property="og:title" content="A new Media Area is online!" />',
            '<meta name="twitter:title" content="A new Media Area is online!" />',
            '<link sizes="16x16" type="image/png" rel="icon" href="https://example.org/favicon.png?h=16&w=16" />',
            '<link sizes="32x32" type="image/png" rel="icon" href="https://example.org/favicon.png?h=32&w=32" />'
        ].join("\n"));
    });
});
