// gatsby-source-datocms/utils.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _require = require('datocms-client'),
    SiteClient = _require.SiteClient,
    Loader = _require.Loader;

var CLIENT_HEADERS = {
  'X-Reason': 'dump',
  'X-SSG': 'gatsby'
};
var GATSBY_CLOUD = process.env.GATSBY_CLOUD;
var GATSBY_EXECUTING_COMMAND = process.env.gatsby_executing_command;
var client, loader;

function createClient(_ref) {
  var apiToken = _ref.apiToken,
      apiUrl = _ref.apiUrl,
      environment = _ref.environment;
  return apiUrl ? new SiteClient(apiToken, _objectSpread({}, CLIENT_HEADERS, {
    environment: environment
  }), apiUrl) : new SiteClient(apiToken, _objectSpread({}, CLIENT_HEADERS, {
    environment: environment
  }));
}

function createLoader(_ref2) {
  var apiToken = _ref2.apiToken,
      apiUrl = _ref2.apiUrl,
      previewMode = _ref2.previewMode,
      environment = _ref2.environment;

  if (!client) {
    client = createClient({
      apiToken: apiToken,
      apiUrl: apiUrl,
      environment: environment
    });
  }

  return new Loader(client, GATSBY_CLOUD && GATSBY_EXECUTING_COMMAND === 'develop' || previewMode, environment);
}

function getLoader(options) {
  if (!loader) {
    loader = createLoader(options);
  }

  return loader;
}

function getClient(options) {
  if (!client) {
    client = createClient(options);
  }

  return client;
}

module.exports = {
  getClient: getClient,
  getLoader: getLoader
};
